admin/cluster_setup.py

```diff
@@ -177,66 +177,64 @@ def main(reactor, args, base_path, top_level):
-                                               node.address,
-                                               remote_logs_file)
-                               )
-        gather_deferreds(results)
-
-        result = 0
-
-    except BaseException:
-        result = 1
-        raise
-    finally:
-        if options['no-keep'] or result == 1:
-            runner.stop_cluster(reactor)
-        else:
-            if cluster is None:
-                print("Didn't finish creating the cluster.")
-                runner.stop_cluster(reactor)
-            else:
-                environment_variables = get_trial_environment(cluster)
-                environment_strings = list()
-                for environment_variable in environment_variables:
-                    environment_strings.append(
-                        "export {name}={value};\n".format(
-                            name=environment_variable,
-                            value=shell_quote(
-                                environment_variables[environment_variable]
-                            ),
-                        )
-                    )
-                environment = ''.join(environment_strings)
-                print("The following variables describe the cluster:")
-                print(environment)
-                env_file = options['cert-directory'].child("environment.env")
-                env_file.setContent(environment)
-                print("The variables are also saved in {}".format(
-                    env_file.path
-                ))
-                print("Be sure to preserve the required files.")
-
-    raise SystemExit(result)
+    results = []
+    yield runner.ensure_keys(reactor)
+    cluster = yield runner.start_cluster(reactor)
+
+    managed_config_file = options['cert-directory'].child("managed.yaml")
+    managed_config = create_managed_config(options['config'], cluster)
+    managed_config_file.setContent(
+        yaml.safe_dump(managed_config, default_flow_style=False)
+    )
+
+    if options['distribution'] in ('centos-7',):
+        remote_logs_file = open("remote_logs.log", "a")
+        for node in cluster.all_nodes:
+            results.append(capture_journal(reactor,
+                                           node.address,
+                                           remote_logs_file)
+                           )
+    elif options['distribution'] in ('ubuntu-14.04', 'ubuntu-15.10'):
+        remote_logs_file = open("remote_logs.log", "a")
+        for node in cluster.all_nodes:
+            results.append(capture_upstart(reactor,
+                                           node.address,
+                                           remote_logs_file)
+                           )
+    # gather_deferreds() below does more than just creating a DeferredList.
+    # So, while we are not using its result in any way, it is still useful
+    # because it logs any failures coming from the gathered deferreds.
+    gather_deferreds(results)
```

Comment:
:question: The idiomatic (and just less confusing) way to do this would be to change the above calls to `capture_upstart` and `capture_journal` to be like `capture_xxx(...).addErrback(write_failure)`.  Then `results` isn't even needed anymore, afaik.  Since this is how it worked before and not really related to this refactoring, optional.